ompi/mca/coll/adapt/coll_adapt_ireduce.c

```diff
@@ -0,0 +1,910 @@
+    /* A free list contains all recv data */
+    opal_free_list_t *inbuf_list;
+    opal_mutex_t *mutex_recv_list;
+    opal_mutex_t *mutex_num_recv_segs;
+    opal_mutex_t *mutex_num_sent;
+    opal_mutex_t **mutex_op_list;
+    /* A list to store the segments need to be sent */
+    opal_list_t *recv_list;
+
+    /* Determine number of segments and number of elements sent per operation */
+    rank = ompi_comm_rank(comm);
+    ompi_datatype_get_extent(dtype, &lower_bound, &extent);
+    ompi_datatype_type_size(dtype, &typelng);
+    COLL_BASE_COMPUTED_SEGCOUNT(seg_size, typelng, seg_count);
+    num_segs = (count + seg_count - 1) / seg_count;
+    segment_increment = (ptrdiff_t) seg_count *extent;
+    ompi_datatype_get_true_extent(dtype, &true_lower_bound, &true_extent);
+    real_seg_size = true_extent + (ptrdiff_t) (seg_count - 1) * extent;
+
+    /* Atomically set up free list */
+    if (NULL == mca_coll_adapt_component.adapt_ireduce_context_free_list) {
+        opal_free_list_t* fl = OBJ_NEW(opal_free_list_t);
+        opal_free_list_init(fl,
+                            sizeof(ompi_coll_adapt_reduce_context_t),
+                            opal_cache_line_size,
+                            OBJ_CLASS(ompi_coll_adapt_reduce_context_t),
+                            0, opal_cache_line_size,
+                            mca_coll_adapt_component.adapt_context_free_list_min,
+                            mca_coll_adapt_component.adapt_context_free_list_max,
+                            mca_coll_adapt_component.adapt_context_free_list_inc,
+                            NULL, 0, NULL, NULL, NULL);
+        if( !OPAL_ATOMIC_COMPARE_EXCHANGE_STRONG_PTR((opal_atomic_intptr_t *)&mca_coll_adapt_component.adapt_ireduce_context_free_list,
+                                                     &(intptr_t){0}, fl) ) {
+            OBJ_RELEASE(fl);
+        }
+    }
+
+    /* If the current process is not leaf */
+    if (tree->tree_nextsize > 0) {
+        inbuf_list = OBJ_NEW(opal_free_list_t);
+        opal_free_list_init(inbuf_list,
+                            sizeof(ompi_coll_adapt_inbuf_t) + real_seg_size,
+                            opal_cache_line_size,
+                            OBJ_CLASS(ompi_coll_adapt_inbuf_t),
+                            0, opal_cache_line_size,
+                            mca_coll_adapt_component.adapt_inbuf_free_list_min,
+                            mca_coll_adapt_component.adapt_inbuf_free_list_max,
+                            mca_coll_adapt_component.adapt_inbuf_free_list_inc,
+                            NULL, 0, NULL, NULL, NULL);
+        /* Set up next_recv_segs */
+        next_recv_segs = (opal_atomic_int32_t *) malloc(sizeof(int32_t) * tree->tree_nextsize);
+        ompi_coll_adapt_inbuf_t *temp_inbuf =
+            (ompi_coll_adapt_inbuf_t *) opal_free_list_wait(inbuf_list);
+        distance = (char *) temp_inbuf->buff - lower_bound - (char *) temp_inbuf;       //address of inbuf->buff to address of inbuf
```

Comment:
It's a trick to avoid allocating the extent of the data. Due to the datatype we know the gap in the beginning will never be touched, so we can artificially move the pointer. There is a lengthy explanation in one of the header files.